craft_application/services/service_factory.py

```diff
@@ -121,7 +288,25 @@ def __getattr__(self, service: str) -> services.AppService:
                 )
             kwargs.setdefault("project", self.project)
 
-        instance: services.AppService = cls(app=self.app, services=self, **kwargs)
+        instance = cls(app=self.app, services=self, **kwargs)
         instance.setup()
-        setattr(self, service, instance)
+        self._services[service] = instance
         return instance
+
+    def __getattr__(self, name: str) -> services.AppService | type[services.AppService]:
+        """Instantiate a service class.
+
+        This allows us to lazy-load only the necessary services whilst still
+        treating them as attributes of our factory in a dynamic manner.
+        For a service (e.g. ``package``, the PackageService instance) that has not
+        been instantiated, this method finds the corresponding class, instantiates
+        it with defaults and any values set using ``set_kwargs``, and stores the
+        instantiated service as an instance attribute, allowing the same service
+        instance to be reused for the entire run of the application.
+        """
+        result = self.get_class(name) if name.endswith("Class") else self.get(name)
+        setattr(self, name, result)
+        return result
+
+
+ServiceFactory.reset()  # Set up default services on import.
```

Comment:
Wouldn't it create a problems like with `Features` which required them to be reset at the specific point of execution ?

Reply:
I don't see how. This sets the default services upon the initial import, but then allows them to be modified or reset at any time. After the initial import the only time this line would run is if someone runs `importlib.reload` on the module.